### ISSUE TITLE
Review and polish the server CLI

### ISSUE THREAD

**Issue by aeschli** [microsoft/vscode#137658]
- [x] add `--help`
- [x] add `--version`
- [x] make a pass througth all other arguments

**Comment by aeschli** [comment]
Server cli:
* `help`, `version`  :heavy_check_mark: 
 
Server setup
*  `host`, `port`, `pick-port`, `socket-path` :heavy_check_mark: 
      - P1: `host` should not default to `0.0.0.0`
      - P3: merge `port` and `pick-port`
      - P3:  improve the name of `socket-path`
*  `connection-token`, `connection-secret` :heavy_check_mark: 
      - P1: Make mandatory,
      - P1: `connection-token` should also be checked when vsda present,
      - Improve `connection-secret` name: `connection-token-file-path`
*  `disable-websocket-compression`
*  `print-ip-address`, `print-startup-performance`
*  `accept-server-license-terms` :heavy_check_mark: 

VS Code options
*  `user-data-dir` :heavy_check_mark: 
*  `disable-telemetry` :heavy_check_mark: 
*  `log`, `logsPath` :heavy_check_mark: 
*  `driver` :heavy_check_mark: 
*  `fileWatcherPolling`
   - rename to `file-watcher-polling`

VS Code web
*  `workspace`, `folder` :eyes: 
*  `enable-sync` 
*  `github-auth`

Extension management: 
*  `list-extensions`, 'install-extension', 'install-builtin-extension', 'uninstall-extension', 'locate-extension, `show-versions`, `category`, `force`, `do-not-sync`, `pre-release` :heavy_check_mark: 
*  `extensions-dir`, `extensions-download-dir`, `builtin-extensions-dir` :heavy_check_mark: 
*  `start-server`: start server and install extensions.
        - Name might be confusing. Needs documentation to clarify it is only needed in combination with `install-extension` :eyes: 

Remote development options:
*  `enable-remote-auto-shutdown`, `remote-auto-shutdown-without-delay`
*  `exec` ??

**Comment by aeschli** [comment]
changes: 221ac5cee35b97db211bf9aaa03961d6260c4970

**Comment by ItalyPaleAle** [comment]
I would like to propose being able to pass at least the most relevant server options as environmental variables, e.g.: `HOST`, `PORT`, `CONNECTION_TOKEN`, `WITHOUT_CONNECTION_TOKEN`, etc.

There are at least 2 scenarios where this would be particularly useful:

1. When running in Docker. The biggest hurdle is that the server binds itself to `127.0.0.1` (or `localhost`) by default, which inside a container makes it un-accessible from anywhere else. In Docker, the app needs to be bound to `0.0.0.0`. With an env var, it's possible to change the default value in the container (by setting `ENV HOST=0.0.0.0` in the Dockerfile) without having to manually pass the argument in the CLI every time (which could lead to user errors).
2. It would allow creating a .env file, which can be useful in a variety of cases, including when running in a container (using `docker run --env-file ...`), when running as a service (a systemd unit could refer an env file with `EnvironmentFile=...`), etc.

(Please let me know if I should move this to another issue)

**Comment by aeschli** [comment]
@ItalyPaleAle Can you elaborate on 'manually pass the argument in the CLI every time'?
Typically, the server just needs to be called once on startup. I would think this is done in the docker file, e.g. as part of the EXEC command. Setting the ENV variables or an env file would be the same amount of work.

I'm no fan of ENV variables as there's a danger that you can inherit one from a parent process without realizing. If you don't clear them, all your child processes will also use them.

Let's create a separate issue for the feature request. I'd suggest we go with the command line parameters for now and see what the user feedback is.

**Comment by ItalyPaleAle** [comment]
> Can you elaborate on 'manually pass the argument in the CLI every time'?

If a user needs to change the ENTRYPOINT, they need to remember passing `--host` (at the very least) or the app won't work. By using env vars, it's possible to set that value as default within the container so if users forget to pass the host, the app still works. (Normally, CLI flags have precedence over ENV vars, if they need to be overridden)

The other situation is systemd units. For example:

```unit
[Unit]
Description=Visual Studio Code Server (Insiders)
​
[Service]
Type=simple
ExecStart=/usr/bin/bash -c '~/.vscode-server-insiders/bin/code-server-insiders --host 127.0.0.1 --port 8000 --connection-token-file ~/.config/vscode-server-insiders/connection-token'
​
[Install]
WantedBy=default.target
```

Systemd units should technically be immutable (imagine a situation where a package, like a DEB, ships with one, and users should not change the unit directly). Without env vars, it's not possible for users to change the config without modifying the unit file itself. With support for env vars, the unit could load an env file (via `EnvironmentFile=`) that users would be able to use to set config options.

> I'm no fan of ENV variables as there's a danger that you can inherit one from a parent process without realizing. If you don't clear them, all your child processes will also use them.

While the pattern of using ENV vars is a common one (see [12 factors](https://12factor.net/config) ) I do understand why in the context of VS Code this could be an issue now. What you're saying is that if the VS Code server is started with a certain environment, then all child processes of the server (including terminals and debugged apps) inherit that environment too? Then yes, using env vars could be a problem. A workaround that could minimize the issue (but not fully remove it) could be to prefix the variables, such as `VSCODE_HOST`.

**Comment by aeschli** [comment]
@ItalyPaleAle Let's create a new issue for an env file / env variable that we can plan for Feb

**Comment by ItalyPaleAle** [comment]
Done. #141333

**Comment by zWingz** [comment]
why remove the arg `folder` and `workspace` ?
I hope to open a default folder by `code-server.sh --folder path/to/folder`

**Comment by aeschli** [comment]
@zWingz Please file a feature request to bring it back as `default-folder` and `default-workspace`.
The folder/workspace to open needs be set in the browser URL: `?folder=...`. Without providing a `folder` / `workspace` in the browser URL an empty window like in the desktop VS Code will be opened. That was the motivation of the change.
